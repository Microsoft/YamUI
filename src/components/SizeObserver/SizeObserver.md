### Notes for use

SizeObserver allows for conditional rendering based on the component's height or width.


### Examples

```js { "props": { "data-description": "basic" } }

const query = {
  small: { maxWidth: 500 },
  large: { minWidth: 501 },
};

const renderAt = (size) => {
  switch (size) {
    case 'large':
      return <div>Large</div>;
    case 'small':
      return <div>Small</div>;
    default:
      return <div>Other</div>;
  }
};

<div>
  <SizeObserver query={query} renderAt={renderAt} />
</div>
```
