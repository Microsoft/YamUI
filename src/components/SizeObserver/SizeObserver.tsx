/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import { BaseComponentProps } from '../../util/BaseComponent/props';

export interface SizeObserverProps extends BaseComponentProps {
  /**
   * An object of `size` keys mapped to height/width queries. Example:
   *
   *     {
   *        small: {
   *          maxWidth: 500,
   *        },
   *        large: {
   *          minWidth: 501,
   *        },
   *     }
   */
  query: SizeQueriesMap;

  /**
   * A render prop which receives the key of the active query.
   */
  renderAt(size?: string): React.ReactNode;
}

export interface SizeObserverState {
  currentSize?: string;
}

export interface SizeQueries {
  maxWidth?: number;
  minWidth?: number;
  maxHeight?: number;
  minHeight?: number;
}

export interface SizeQueriesMap {
  [key: string]: SizeQueries;
}

const getMatchingQuerySize = (query: SizeQueriesMap, height: number, width: number) => {
  for (const key of Object.keys(query)) {
    const { maxHeight = Infinity, maxWidth = Infinity, minHeight = 0, minWidth = 0 } = query[key];

    if (height <= maxHeight && height >= minHeight && width <= maxWidth && width >= minWidth) {
      return key;
    }
  }
};

/**
 * `SizeObserver` uses the the `ResizeObserver` API to allow conditional child rendering based on the component's height
 * or width. Provide a `renderAt` prop function to render your responsive content based on the given size. For
 * performance, this function will only be called when the size changes.
 */
export default class SizeObserver extends React.Component<SizeObserverProps, SizeObserverState> {
  private observedElement: React.RefObject<HTMLDivElement>;
  private resizeObserver: ResizeObserver;

  constructor(props: SizeObserverProps) {
    super(props);

    this.state = {};

    this.observedElement = React.createRef();
    this.resizeObserver = new ResizeObserver(this.onSizeChange);
  }

  public componentDidMount() {
    this.observe();
  }

  public componentWillUnmount() {
    this.disconnect();
  }

  public render() {
    return <div ref={this.observedElement}>{this.getObserverChildren()}</div>;
  }

  private getObserverChildren() {
    const { currentSize } = this.state;

    return this.props.renderAt(currentSize);
  }

  private onSizeChange = ([entry]: ResizeObserverEntry[]) => {
    const { height, width } = entry.contentRect;
    const size = getMatchingQuerySize(this.props.query, height, width);

    if (this.state.currentSize === size) {
      return;
    }

    this.setState(() => ({ currentSize: size }));
  };

  private observe() {
    if (!this.observedElement.current) {
      return;
    }
    this.resizeObserver.observe(this.observedElement.current);
  }

  private disconnect() {
    this.resizeObserver.disconnect();
  }
}
