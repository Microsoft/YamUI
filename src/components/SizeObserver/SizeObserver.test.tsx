/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import * as React from 'react';
import { mount, ReactWrapper } from 'enzyme';
import SizeObserver, { SizeObserverProps } from './index';

describe('<SizeObserver />', () => {
  let component: ReactWrapper<SizeObserverProps>;
  let onSizeChange: ResizeObserverCallback;

  beforeEach(() => {
    // tslint:disable-next-line:no-function-expression
    (window as any).ResizeObserver = function(callback: ResizeObserverCallback) {
      onSizeChange = callback;
    };
    ResizeObserver.prototype.disconnect = jest.fn();
    ResizeObserver.prototype.observe = jest.fn();
    ResizeObserver.prototype.unobserve = jest.fn();

    const query = {
      small: {
        maxWidth: 500,
        maxHeight: 500,
      },
      large: {
        minWidth: 501,
        minHeight: 501,
      },
    };

    const renderAt = jest.fn((size?: string) => {
      switch (size) {
        case 'large':
          return <span>Large</span>;
        case 'small':
          return <span>Small</span>;
        default:
          return <span>Other</span>;
      }
    });

    component = mount(<SizeObserver query={query} renderAt={renderAt} />);
  });

  it('observes the element', () => {
    expect(window.ResizeObserver.prototype.observe).toHaveBeenCalled();
  });

  it('renders at the default size on first render', () => {
    expect(component.find('span')).toMatchSnapshot();
  });

  describe('when its size changes', () => {
    beforeEach(() => {
      const entries: ResizeObserverEntry[] = [
        {
          target: component.getDOMNode(),
          contentRect: {
            height: 500,
            width: 500,
          } as any,
        },
      ];
      onSizeChange(entries, {} as any);
    });

    it('renders at the corresponding size', () => {
      expect(component.update().find('span')).toMatchSnapshot();
    });
  });

  describe('when its size changes twice but both changes match the same query', () => {
    beforeEach(() => {
      const entries: ResizeObserverEntry[] = [
        {
          target: component.getDOMNode(),
          contentRect: {
            height: 501,
            width: 501,
          } as any,
        },
      ];
      onSizeChange(entries, {} as any);
      onSizeChange(entries, {} as any);
    });

    it('renders at the corresponding size', () => {
      expect(component.update().find('span')).toMatchSnapshot();
    });

    it('calls renderAt once for initialize and once for the first size change', () => {
      expect(component.prop('renderAt')).toHaveBeenCalledTimes(2);
    });
  });

  describe('when unmounted', () => {
    beforeEach(() => {
      component.unmount();
    });

    it('disconnects', () => {
      expect(window.ResizeObserver.prototype.disconnect).toHaveBeenCalled();
    });
  });
});
