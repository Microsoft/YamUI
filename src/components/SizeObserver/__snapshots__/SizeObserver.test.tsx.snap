// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SizeObserver /> renders at the default size on first render 1`] = `
<span>
  Other
</span>
`;

exports[`<SizeObserver /> when its size changes renders at the corresponding size 1`] = `
<span>
  Small
</span>
`;

exports[`<SizeObserver /> when its size changes twice but both changes match the same query renders at the corresponding size 1`] = `
<span>
  Large
</span>
`;
